### ISSUE TITLE
Check/uncheck All and generation of forms

### ISSUE THREAD

**Issue by starapple2** [jan-vandenberg/cruddiy#123]
A couple of matters of clarification: 

1. I am generating pages primarily for the adding of new records and I see the two check boxes to Check/uncheck all columns in a table. What is the purpose of the second of the two? When I click on the first box it toggles the selection of all columns. I don't grasp what the second one does.
 ![check-uncheck-Columns](https://github.com/jan-vandenberg/cruddiy/assets/33574205/37c9176f-ed23-49af-b4eb-d7b17d3c4090)
2. My second query is with regards limiting what columns are displayed as fields in the generated forms, which I thought was affected by the _Show column_ function.

The tables I am working with have fields that are restricted to back-end (admin) users after forms are submitted by public-facing users. It seems I would have to edit the generated forms to prevent the showing of fields such as table IDs, account status, time stamps or text areas for admin notes.

In summary what is the purpose of the second toggle and can one decide what columns are generated as form fields? The completed app from the image above would correctly show an HTML table without the account_id column but the account-create.php would display the account_id field.

Is the ID field set depending on whether the column is auto incremented, because the behaviour isn't consistent with all column IDs? The setting here:
![image](https://github.com/jan-vandenberg/cruddiy/assets/33574205/05b1077e-b9e8-4a3d-9226-9fb9f29c6619)

Results in (although other columns not selected also are in the form):
![image](https://github.com/jan-vandenberg/cruddiy/assets/33574205/af7ac88f-deb2-40e5-8754-5d18af7eefff)

**Comment by germain-italic** [comment]
> 1. What is the purpose of the second of the two?

You have spotted a glitch in UI, the second column of checkboxes should be only visible when at least one table has a foreign key.

The checkboxes under "This table" are used to display the selected column in the list view of each table.

If you had foreign keys in your schema, you would see "Related tables" in column #2.
It would be used to display the selected column when it is referenced as foreign key in another table. Columns keys id, name, reference are checked by default.

Here's an example:

![](https://files.italic.fr/chrome_k7eg9oR41R.png)

I have some ideas on how to improve this, will implement in the coming weeks.

**Comment by germain-italic** [comment]
> 2. the Show column function

It changes the layout of the table onto the generated "Read" pages (_xxxx-index.php_). You should already have:

> The completed app from the image above would correctly show an HTML table without the account_id column

I've checked the code and I haven't seen any impact on the Create/Edit pages so far.
That's a cool idea to implement.

In order to avoid breaking compatibility with existing apps, I suggest that we offer the choice.
Sometimes you want a compact table, but a full-fledged form.
My suggestion:

![](https://files.italic.fr/chrome_FJFDQ5UVRi.png)

**Does it make sense?
Is the wording clear enough?**

We've struggled to find a presentation for this concept (see [this comment](https://github.com/jan-vandenberg/cruddiy/pull/64#issuecomment-1568610889) and below) so any suggestion is welcome.

> Is the ID field set depending on whether the column is auto incremented

- The ID will be auto incremented if your table schema says so (ex: `CREATE TABLE `brands` (`id` int(11) NOT NULL AUTO_INCREMENT`). 
- If you've set NOT NULL without auto-increment it will produce an error.

The ID is passed via GET so that won't make any difference if we remove the field from the form.

We also need to hide the unchecked columns on _xxxx-read.php_ and _xxxx-update.php_, right?

**Comment by starapple2** [comment]
> > 1. What is the purpose of the second of the two?
> 
> You have spotted a glitch in UI, the second column of checkboxes should be only visible when at least one table has a foreign key.
> 
> The checkboxes under "This table" are used to display the selected column in the list view of each table.
> 
> If you had foreign keys in your schema, you would see "Related tables" in column #2. It would be used to display the selected column when it is referenced as foreign key in another table. Columns keys id, name, reference are checked by default.
> 
> Here's an example:
> 
> ![](https://camo.githubusercontent.com/3c3d215e8b6299b65250edc8de45ae19e36d4ded0b65b4322add8394850d3cd1/68747470733a2f2f66696c65732e6974616c69632e66722f6368726f6d655f6b376567396f523431522e706e67)
> 
> _This will show a `countries.name` column on _users-index.php_ (without `countries.id` column)_
> 
> I have some ideas on how to improve this, will implement in the coming weeks.

I suspected as much that it was related to the presence of foreign keys. The original db as MyIsaam and the two columns of and I thought after reading the user guide that that should not be the case if the tables were InnoDB; so I changed them to InnoDB and the behaviour didn't change.

I think we're on the same page with your 2. re showing/hiding.

Best.

**Comment by germain-italic** [comment]
Do you _actually_ have a foreign key?

Does the relation appears (or has been created) on in step2?

![](https://files.italic.fr/chrome_OvUmKW18Fo.png)

**Comment by michielsebastiaan** [comment]
I have a question in addition to this:
Would it be possible to set some fields to "read only" (Not having the crud options?)
So, for instance some values may only be set once (When creating a new entry)
